govcd/edgegateway_lb_service_monitor.go

```diff
@@ -0,0 +1,187 @@
+					monitor.Name, monitor.ID, lbMonitorConfig.ID)
+			}
+			return monitor, nil
+		}
+	}
+
+	return nil, fmt.Errorf("could not find load balancer service monitor (name: %s, ID: %s)",
+		lbMonitorConfig.ID, lbMonitorConfig.Name)
+}
+
+// UpdateLBServiceMonitor
+func (eGW *EdgeGateway) UpdateLBServiceMonitor(lbMonitorConfig *types.LBMonitor) (*types.LBMonitor, error) {
+	if err := validateUpdateLBServiceMonitor(lbMonitorConfig); err != nil {
+		return nil, err
+	}
+
+	// if only name was specified for update, ID must be found, because ID is mandatory for update
+	if lbMonitorConfig.ID == "" {
+		readLBMonitor, err := eGW.ReadLBServiceMonitor(&types.LBMonitor{Name: lbMonitorConfig.Name})
+		if err != nil {
+			return nil, fmt.Errorf("unable to find load balancer monitor by name for update: %s", err)
+		}
+		lbMonitorConfig.ID = readLBMonitor.ID
+	}
+
+	httpPath, err := eGW.getProxiedEdgeEndpoint("/loadbalancer/config/monitors/" + lbMonitorConfig.ID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get Edge Gateway API endpoint: %s", err)
+	}
+
+	// Result should be 204, if not we expect an error of type types.NSXError
+	_, err = eGW.client.ExecuteRequestHTTPCodeOrTypedError(http.StatusNoContent, httpPath, http.MethodPut, "application/xml", "%s", lbMonitorConfig, &types.NSXError{})
+	if err != nil {
+		return nil, fmt.Errorf("error while updating load balancer service monitor : %s", err)
+	}
+
+	readMonitor, err := eGW.ReadLBServiceMonitor(&types.LBMonitor{ID: lbMonitorConfig.ID})
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve monitor with ID (%s) after update: %s", readMonitor.ID, err)
+	}
+	return readMonitor, nil
+}
+
+// DeleteLBServiceMonitor is able to delete the types.LBMonitor type by Name and/or ID.
+// If both - Name and ID are specified it performs a lookup by ID and returns an error if the specified name and found
+// name do not match.
+func (eGW *EdgeGateway) DeleteLBServiceMonitor(lbMonitorConfig *types.LBMonitor) error {
+	if err := validateDeleteLBServiceMonitor(lbMonitorConfig); err != nil {
+		return err
+	}
+
+	lbMonitorID := lbMonitorConfig.ID
+	// if only name was specified for deletion, ID must be found, because only ID can be used for deletion
+	if lbMonitorConfig.ID == "" {
+		readLBMonitor, err := eGW.ReadLBServiceMonitor(&types.LBMonitor{Name: lbMonitorConfig.Name})
+		if err != nil {
+			return fmt.Errorf("unable to find load balancer monitor by name for deletion: %s", err)
+		}
+		lbMonitorID = readLBMonitor.ID
+	}
+
+	httpPath, err := eGW.getProxiedEdgeEndpoint("/loadbalancer/config/monitors/" + lbMonitorID)
+	if err != nil {
+		return fmt.Errorf("could not get Edge Gateway API endpoint: %s", err)
+	}
+	return eGW.client.ExecuteRequestWithoutResponse(httpPath, http.MethodDelete, "application/xml", "unable to delete Service Monitor: %s", nil)
+}
+
+func validateCreateLBServiceMonitor(lbMonitorConfig *types.LBMonitor) error {
```

Comment:
Do we need to export these validation functions to use with Terraform?

Reply:
We don't necessarily as they're used inside CRUD methods (`CreateLBServiceMonitor`, `ReadLBServiceMonitor`, etc) implicitly. I guess it can stay like that now.